### PR TITLE
[3.2] Fix incorrect serializing of std::optional when value is not provided - (GH #1189)

### PR DESCRIPTION
**Pull request by greg7mdp**
When serializing a `std::optional` field, and a value is not provided for the `std::optional`, we need to add into the serialization the flag specifying that a missing value is serialized. Prior to this PR, the flag was missing.

### CODE REVIEW

File: libraries/libfc/include/fc/time.hpp [hunk 1]
```diff
@@ -11,7 +12,7 @@ namespace fc {
   class microseconds {
     public:
         constexpr explicit microseconds( int64_t c = 0) :_count(c){}
-        static constexpr microseconds maximum() { return microseconds(0x7fffffffffffffffll); }
+        static constexpr microseconds maximum() { return microseconds(std::numeric_limits<int64_t>::max()); }
```
Comment (heifner) [inline libraries/libfc/include/fc/time.hpp:15]: This seems like unnecessary changes for a backport. I would suggest we only do this for `main`.

File: unittests/abi_tests.cpp [hunk 2]
```diff
@@ -40,7 +40,17 @@ FC_REFLECT(act_sig, (sig) )
 
 BOOST_AUTO_TEST_SUITE(abi_tests)
 
+#ifdef NDEBUG
 fc::microseconds max_serialization_time = fc::seconds(1); // some test machines are very slow
+#else
+fc::microseconds max_serialization_time = fc::microseconds::maximum(); // don't check in debug builds
+#endif
```
Comment (heifner) [inline unittests/abi_tests.cpp:47]: Was this because it was failing locally on a debug build? I think just increasing the existing to 2 or 3 seconds would be preferred to having diff behavior for debug.

**Review by heifner — CHANGES_REQUESTED** [review]

File: unittests/abi_tests.cpp (hunk 2 above)
Comment (greg7mdp) [reply unittests/abi_tests.cpp:47]: No, it was because the timeouts were hit when I was stepping through the code. The debug build purpose is for debugging, so having timeouts is unfriendly for that). Is there a reason why timeouts in abi tests would be useful in debug builds?

File: libraries/libfc/include/fc/time.hpp (hunk 1 above)
Comment (greg7mdp) [reply libraries/libfc/include/fc/time.hpp:15]: sure, will remove for the backports.

File: unittests/abi_tests.cpp (hunk 2 above)
Comment (heifner) [reply unittests/abi_tests.cpp:47]: Ah, that makes sense. There is a potential for a hang with a debug build, but one could always attach a debugger in that case.

**Review by heifner — APPROVED** [review]

**Comment by greg7mdp** [comment]
@heifner @arhag One concern I have is that this PR changes the serialization output for structs containing `std::optional`, when a value is omitted for one or more `std::optional`.
My first thought is that it may not be an issue, because the deserialization of these would have failed anyways, but I thought it would be worth mentioning.

**Comment by heifner** [comment]
> @heifner @arhag One concern I have is that this PR changes the serialization output for structs containing `std::optional`, when a value is omitted for one or more `std::optional`. My first thought is that it may not be an issue, because the deserialization of these would have failed anyways, but I thought it would be worth mentioning.

I don't anticipate any issue here.

File: unittests/abi_tests.cpp (hunk 2 above)
Comment (linh2931) [reply unittests/abi_tests.cpp:47]: It might be helpful to add a comment explaining in the test code why #if NDEBUG is required.

File: unittests/abi_tests.cpp [hunk 3]
```diff
@@ -1929,6 +1937,113 @@ BOOST_AUTO_TEST_CASE(abi_type_loop)
...
+        "bridge_fee" : "2"
+      }
+      )=====";
+
+      auto var = fc::json::from_string(test_data);
+      verify_byte_round_trip_conversion(abis, "fees", var);
+   }
+
+   {
+      // check conversion when the first optional members is missing
```
Comment (linh2931) [inline unittests/abi_tests.cpp:2009]: `first` should be `second` in the comment.

File: unittests/abi_tests.cpp [hunk 4]
```diff
@@ -1929,6 +1937,113 @@ BOOST_AUTO_TEST_CASE(abi_type_loop)
...
+        "bridge_fee" : "2"
+      }
+      )=====";
+
+      auto var = fc::json::from_string(test_data);
+      verify_byte_round_trip_conversion(abis, "fees", var);
+   }
+
+   {
+      // check conversion when all optional members are provided
```
Comment (linh2931) [inline unittests/abi_tests.cpp:2022]: should be last.

File: unittests/abi_tests.cpp [hunk 5]
```diff
@@ -1929,6 +1937,113 @@ BOOST_AUTO_TEST_CASE(abi_type_loop)
...
+        "miner_cut" : "2",
+      }
+      )=====";
+
+      auto var = fc::json::from_string(test_data);
+      verify_byte_round_trip_conversion(abis, "fees", var);
+   }
+
+   {
+      // check conversion when all optional members are provided
```
Comment (linh2931) [inline unittests/abi_tests.cpp:2035]: should be none.

**Review by linh2931 — CHANGES_REQUESTED** [review]

File: unittests/abi_tests.cpp (hunk 5 above)
Comment (greg7mdp) [reply unittests/abi_tests.cpp]: Thanks, Lin, I updated all the comments.

**Review by linh2931 — APPROVED** [review]
